### PR TITLE
RFC: Explicit init command

### PR DESCRIPTION
**Pull request by bdarnell**
This RFC proposes a new init command to be run after nodes have been
started, to take the place of automatic bootstrapping on a node with
no join flags.

CC @spencerkimball @petermattis @a-robinson @mberhault

---
This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/14251)

### CODE REVIEW

**Comment by petermattis** [comment]
I haven't thought this through sufficiently, but another approach struck me today: before calling `start` the first time, we dump an empty marker file (`BOOTSTRAP`) on the node we wish to bootstrap the cluster. This avoids worries about RPC failures and how to handle that situation. An empty marker file seems easier to script than running `init` specially on one machine, though I'm not familiar enough with kubernetes and docker swarm to know if this would be better in practice.

---

Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[docs/RFCS/init_command.md, line 119 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14251#-KfZ7eG9vsdyngE277jq:-KfZ7eGAoXuRdHrxso_Z:b-px02s) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab226028f90995c1e32f628b7636688f54ac5cc/docs/RFCS/init_command.md#L119)):*
> ```Markdown
> - Should we remove/deprecate the auto-bootstrapping when no `--join`
>   flag is present so that there is only one way to do it, or keep it
>   as a simpler option for some environments?
> ```

I'd hate to lose this. I run `rm -fr cockroach-data && ./cockroach start` all the time. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14251#-:-KfZ6kWu2Jo8jBTC4u62:b-an49k3)*

**Comment by bdarnell** [comment]
I don't think that dropping a marker file is significantly easier in practice than running `init` (as evidence, see our kubernetes init container, which does exactly this: it drops a marker file which the main container will use to decide whether or not to pass `--join`). The problem is that the sequencing capabilities provided by these tools are fairly limited, as are the tools for working with a disk allocation not associated with a running process. You can make it work, but you end up combining a lot of rarely-used parts of the system to get there.

---

Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[docs/RFCS/init_command.md, line 119 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14251#-KfZ7eG9vsdyngE277jq:-KfZBC7AmPFb18iyP8JU:b71v90y) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab226028f90995c1e32f628b7636688f54ac5cc/docs/RFCS/init_command.md#L119)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I'd hate to lose this. I run `rm -fr cockroach-data && ./cockroach start` all the time. 
</blockquote></details>

What if we (re-)added `cockroach start --init` to perform the initialization in one step?  (or `--singlenode`, which would be a combination of `init`, `start`, and change the zone config to use a single replica and quiet the logs? I'd even go so far as to suggest that a single-node server run this way should run a sql shell on its stdin/stdout)

We're also fairly unusual in that all our scripts start out by doing `CREATE (DATABASE|TABLE) IF NOT EXISTS`. Most apps in my experience have an explicit database setup phase into which the `cockroach init` command would fit fairly nicely. (Most people would also use DROP DATABASE instead of blowing away the data directory completely. Our use cases as developers of the database are not the same as those of our users)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14251#-:-KfZ9y8EFw8GKrxHAoGN:b-d3pds5)*

**Comment by mberhault** [comment]
Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[docs/RFCS/init_command.md, line 109 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14251#-Kf_yzD6fEWGGkLNAUe-:-Kf_yzD6fEWGGkLNAUe0:b7xg8cz) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab226028f90995c1e32f628b7636688f54ac5cc/docs/RFCS/init_command.md#L109)):*
> ```Markdown
> didn't take the number of expected nodes, and simply made a single
> attempt to call `InitCluster`. If this failed with a network error,
> the admin would be responsible for determining whether the
> ```

Given that this is run directly after starting the nodes, it seems more likely that errors would be due to misconfiguration (port firewalled, hostname not in server certs) than temporary errors (although some are definitely possible, including name resolution delay).
If the init command is pointed at a specific node, I don't think counting is necessary. Pointing it at a load-balancer (or other one-to-many) would make it required, but then we also have to worry about forwarding (to make sure we always hit the right node) and health endpoints returning OK during the bootstrap phase.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14251)*

**Comment by benesch** [comment]
Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[docs/RFCS/init_command.md, line 119 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14251#-KfZ7eG9vsdyngE277jq:-KfbQT_I6PPFzLHREsdJ:b-5kgcw1) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab226028f90995c1e32f628b7636688f54ac5cc/docs/RFCS/init_command.md#L119)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

What if we (re-)added `cockroach start --init` to perform the initialization in one step?  (or `--singlenode`, which would be a combination of `init`, `start`, and change the zone config to use a single replica and quiet the logs? I'd even go so far as to suggest that a single-node server run this way should run a sql shell on its stdin/stdout)

We're also fairly unusual in that all our scripts start out by doing `CREATE (DATABASE|TABLE) IF NOT EXISTS`. Most apps in my experience have an explicit database setup phase into which the `cockroach init` command would fit fairly nicely. (Most people would also use DROP DATABASE instead of blowing away the data directory completely. Our use cases as developers of the database are not the same as those of our users)
</blockquote></details>

We've also got `make start`, which could learn to run whatever commands are necessary under the new design. It seems like the need for "spin up a single, bootstrapped, non-daemonized node" is confined to Cockroach developers. I imagine most users who need a local Cockroach install for development will use a Homebrew/systemd service once that's available. (I've certainly never run `mysqld` or `postmaster` myself.) 

Plus, if you're trying CockroachDB for the first time and don't have an active install, our cluster quick start currently recommends this mouthful:

```
cockroach start --background
cockroach start --background --store=node2 --port=26258 --http-port=8081 --join=localhost:26257
cockroach start --background --store=node3 --port=26259 --http-port=8082 --join=localhost:26258
```

Having to run `cockroach init` at the end of that sequence doesn't strike me as particularly onerous.

So tl;dr I'm in favor of dropping auto-bootstrap and providing facilities for launching convenient single-node clusters through `make start` (for Cockroach developers) and OS services (for users).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14251)*

**Comment by a-robinson** [comment]
In Kubernetes, it would be a bit tougher to ensure that the empty marker file gets written just once (rather than also getting rewritten when the first instance loses its disk for whatever reason). It'd end up looking pretty much exactly like what we have today (except that we'd be touching a file rather than modifying command-line flags).

The safest way to run an init command only once against the cluster will require the user to do it themselves (rather than burying it into the same config file), which doesn't seem ideal from a UX perspective, but it would be less fragile and remove a bunch of the mess from the configuration.

The init command would be much nicer for swarm, since we're already requiring an extra command to be run there in order to change the join flags after initialization.

Have you checked out prior art on this? What do other similar systems do?

---

Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[docs/RFCS/init_command.md, line 5 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14251#-KfljA18wyMzTqctl-1I:-KfljA19p6tL0A2Xzmvs:b-bk7zoh) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab226028f90995c1e32f628b7636688f54ac5cc/docs/RFCS/init_command.md#L5)):*
> ```Markdown
> - Start Date: 2017-03-13
> - Authors: @bdarnell
> - RFC PR: (PR # after acceptance of initial draft)
> ```

Might as well update this on your next revision

---

*[docs/RFCS/init_command.md, line 6 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14251#-KfljFwGdtysTISbo8hq:-KfljFwGdtysTISbo8hr:bluifga) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab226028f90995c1e32f628b7636688f54ac5cc/docs/RFCS/init_command.md#L6)):*
> ```Markdown
> - Authors: @bdarnell
> - RFC PR: (PR # after acceptance of initial draft)
> - Cockroach Issue:
> ```

[#5974](https://github.com/cockroachdb/cockroach/issues/5974)

---

*[docs/RFCS/init_command.md, line 62 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14251#-KfljdnV_QKEwCn0pyO1:-KfljdnV_QKEwCn0pyO2:b-6isgky) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab226028f90995c1e32f628b7636688f54ac5cc/docs/RFCS/init_command.md#L62)):*
> It polls gossip via the
> status endpoint until that number of nodes are present

I expect we'll also want to let it be run without requiring a specific number of nodes.

Also, what happens if the cluster has more nodes than are specified in the command? Success, I guess?

---

*[docs/RFCS/init_command.md, line 109 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14251#-Kf_yzD6fEWGGkLNAUe-:-KflkSvxAxEYNcpnargD:b-euyex6) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab226028f90995c1e32f628b7636688f54ac5cc/docs/RFCS/init_command.md#L109)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

Given that this is run directly after starting the nodes, it seems more likely that errors would be due to misconfiguration (port firewalled, hostname not in server certs) than temporary errors (although some are definitely possible, including name resolution delay).
If the init command is pointed at a specific node, I don't think counting is necessary. Pointing it at a load-balancer (or other one-to-many) would make it required, but then we also have to worry about forwarding (to make sure we always hit the right node) and health endpoints returning OK during the bootstrap phase.
</blockquote></details>

+1 to all of this.

---

*[docs/RFCS/init_command.md, line 119 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14251#-KfZ7eG9vsdyngE277jq:-KflkzbIjQUEXpPU2N3J:b8935zg) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab226028f90995c1e32f628b7636688f54ac5cc/docs/RFCS/init_command.md#L119)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

We've also got `make start`, which could learn to run whatever commands are necessary under the new design. It seems like the need for "spin up a single, bootstrapped, non-daemonized node" is confined to Cockroach developers. I imagine most users who need a local Cockroach install for development will use a Homebrew/systemd service once that's available. (I've certainly never run `mysqld` or `postmaster` myself.) 

Plus, if you're trying CockroachDB for the first time and don't have an active install, our cluster quick start currently recommends this mouthful:

```
cockroach start --background
cockroach start --background --store=node2 --port=26258 --http-port=8081 --join=localhost:26257
cockroach start --background --store=node3 --port=26259 --http-port=8082 --join=localhost:26258
```

Having to run `cockroach init` at the end of that sequence doesn't strike me as particularly onerous.

So tl;dr I'm in favor of dropping auto-bootstrap and providing facilities for launching convenient single-node clusters through `make start` (for Cockroach developers) and OS services (for users).
</blockquote></details>

I like the idea of having an `--init` flag on cockroach start that will basically maintain today's behavior of doing bootstrapping if the store directory is empty. It'll continue supporting all existing workflows, which are very nice not just for us but also for people that simply want to run a command (or 3 for a cluster) and have a functioning database to kick the tires with.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14251#-:-Kfl9lNipDsAHiLlNXev:blk8s9a)*

**Comment by benesch** [comment]
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[docs/RFCS/init_command.md, line 119 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14251#-KfZ7eG9vsdyngE277jq:-KfnIw2Rw2bh6YqAuPb5:b-58t27n) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab226028f90995c1e32f628b7636688f54ac5cc/docs/RFCS/init_command.md#L119)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

I like the idea of having an `--init` flag on cockroach start that will basically maintain today's behavior of doing bootstrapping if the store directory is empty. It'll continue supporting all existing workflows, which are very nice not just for us but also for people that simply want to run a command (or 3 for a cluster) and have a functioning database to kick the tires with.
</blockquote></details>

We might be better served by a `cockroach start-dev-cluster --nodes=N` command. I think we're trying to shoehorn too much into `start`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14251)*

**Comment by bdarnell** [comment]
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[docs/RFCS/init_command.md, line 5 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14251#-KfljA18wyMzTqctl-1I:-KfrgyOkMXndFM5xJvzn:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab226028f90995c1e32f628b7636688f54ac5cc/docs/RFCS/init_command.md#L5)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

Might as well update this on your next revision
</blockquote></details>

Done.

---

*[docs/RFCS/init_command.md, line 6 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14251#-KfljFwGdtysTISbo8hq:-KfrgygUsnYlU7YqVtNo:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab226028f90995c1e32f628b7636688f54ac5cc/docs/RFCS/init_command.md#L6)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

[#5974](https://github.com/cockroachdb/cockroach/issues/5974)
</blockquote></details>

Done.

---

*[docs/RFCS/init_command.md, line 62 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14251#-KfljdnV_QKEwCn0pyO1:-KfrgzNWKvPTKSUyRccw:b-e3hqsn) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab226028f90995c1e32f628b7636688f54ac5cc/docs/RFCS/init_command.md#L62)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

> It polls gossip via the
> status endpoint until that number of nodes are present

I expect we'll also want to let it be run without requiring a specific number of nodes.

Also, what happens if the cluster has more nodes than are specified in the command? Success, I guess?
</blockquote></details>

I was thinking it would be an error if the number of nodes didn't match, in either direction. 

We could make the number of nodes optional at the expense of occasionally making some failure situations ambiguous.

---

*[docs/RFCS/init_command.md, line 109 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14251#-Kf_yzD6fEWGGkLNAUe-:-KfrhTLrz05SzbbbzFNm:btrmg25) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab226028f90995c1e32f628b7636688f54ac5cc/docs/RFCS/init_command.md#L109)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

+1 to all of this.
</blockquote></details>

OK, that's two votes, does anyone want to keep the node-count verification logic or shall I just simplify this to the 

---

*[docs/RFCS/init_command.md, line 119 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14251#-KfZ7eG9vsdyngE277jq:-KfrhjzgGMLmjct9VAW2:byjtwxn) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab226028f90995c1e32f628b7636688f54ac5cc/docs/RFCS/init_command.md#L119)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

We might be better served by a `cockroach start-dev-cluster --nodes=N` command. I think we're trying to shoehorn too much into `start`.
</blockquote></details>

My thinking now (after some offline conversation) is that for backwards compatibility we should leave the existing behavior unchanged.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14251)*

**Comment by mberhault** [comment]
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[docs/RFCS/init_command.md, line 109 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14251#-Kf_yzD6fEWGGkLNAUe-:-KfuQvh1Cc8QnC4OJb2T:bfpdmtf) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab226028f90995c1e32f628b7636688f54ac5cc/docs/RFCS/init_command.md#L109)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

OK, that's two votes, does anyone want to keep the node-count verification logic or shall I just simplify this to the 
</blockquote></details>

I'm not sure we're voting against node-counting. And I probably put my comment in the wrong place. This belongs in the `detailed design` section.

Let me summarize two scenarios, both likely to be common deployment strategies:
1. all nodes are directly addressable at creation time. eg: `node-1`, `node-2`, ... `node-N`. init with `cockroach init --host=node-1`
2. nodes are not addressable externally, but can be reached through `node-lb` which randomly forwards request to one of the nodes. init with`cockroach init --host=node-lb`

The first case is easy as we're guaranteed to always talk to the same node. Counting is a nicety here, providing some feedback but not strictly necessary.

The second case is tricker. Some of the init steps are as described in the detailed design, but not all (highlighted):
* __respond OK on health pages to allow load balancer to server to all nodes__ (we could require tcp-level health checks, but that's not always an option)
* poll gossip status page until we hit the number of nodes. it doesn't matter which since we're looking for exactly the expected number of nodes. Either all nodes have the same gossip page, or we don't have the right number of nodes.
* pick a node. lexicographically should be fine, or at least I can't think of a reason why it would be a problem.
* issue `InitCluster` RPC, __with "forward-to: picked node address"__. We do have forwarding in status pages, but those are by node ID. Forwarding by address is even easier.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14251)*

**Comment by a-robinson** [comment]
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[docs/RFCS/init_command.md, line 62 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14251#-KfljdnV_QKEwCn0pyO1:-KfvQ_F_o2LCIYbNnJHs:b-nwspxi) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab226028f90995c1e32f628b7636688f54ac5cc/docs/RFCS/init_command.md#L62)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I was thinking it would be an error if the number of nodes didn't match, in either direction. 

We could make the number of nodes optional at the expense of occasionally making some failure situations ambiguous.
</blockquote></details>

The good thing is that `cockroach init` will only be run when the cluster isn't holding any user data and (as long as we're careful about retries) its problems will usually occur when the cluster is misconfigured and would require some manual poking and prodding anyways.

---

*[docs/RFCS/init_command.md, line 119 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14251#-KfZ7eG9vsdyngE277jq:-KfvRIqAMwhY8txXwE1q:b44tpx) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab226028f90995c1e32f628b7636688f54ac5cc/docs/RFCS/init_command.md#L119)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

My thinking now (after some offline conversation) is that for backwards compatibility we should leave the existing behavior unchanged.
</blockquote></details>

👍

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14251)*

**Comment by a-robinson** [comment]
Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14251)*

File: docs/RFCS/init_command.md [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+introducing a `cockroach init` command which can take the place of the
+current logic involving the absence of a `--join` flag. This is
+intended to be more compatible with various deployment tools by making
+the node configuration more homogeneous.
+
+The new procedure will be:
+
+1. Start all nodes with the same `--join` flag.
+2. Run `cockroach init --host=... N`, where `N` is the number of nodes started.
+
```
Comment (vivekmenezes) [inline docs/RFCS/init_command.md]: An example using a three node cluster would be nice here.

**Comment by bdarnell** [comment]
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[docs/RFCS/init_command.md, line 109 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14251#-Kf_yzD6fEWGGkLNAUe-:-KfwUTMlrFpnQDVOPVTQ:b-gptg4f) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab226028f90995c1e32f628b7636688f54ac5cc/docs/RFCS/init_command.md#L109)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

I'm not sure we're voting against node-counting. And I probably put my comment in the wrong place. This belongs in the `detailed design` section.

Let me summarize two scenarios, both likely to be common deployment strategies:
1. all nodes are directly addressable at creation time. eg: `node-1`, `node-2`, ... `node-N`. init with `cockroach init --host=node-1`
2. nodes are not addressable externally, but can be reached through `node-lb` which randomly forwards request to one of the nodes. init with`cockroach init --host=node-lb`

The first case is easy as we're guaranteed to always talk to the same node. Counting is a nicety here, providing some feedback but not strictly necessary.

The second case is tricker. Some of the init steps are as described in the detailed design, but not all (highlighted):
* __respond OK on health pages to allow load balancer to server to all nodes__ (we could require tcp-level health checks, but that's not always an option)
* poll gossip status page until we hit the number of nodes. it doesn't matter which since we're looking for exactly the expected number of nodes. Either all nodes have the same gossip page, or we don't have the right number of nodes.
* pick a node. lexicographically should be fine, or at least I can't think of a reason why it would be a problem.
* issue `InitCluster` RPC, __with "forward-to: picked node address"__. We do have forwarding in status pages, but those are by node ID. Forwarding by address is even easier.
</blockquote></details>

Yeah, I was thinking that even with node-counting, we would require that the `init` command be issued directly to a node. I don't want to build in a forwarding mechanism (at this time). 

The point of counting nodes is to eliminate ambiguous results, enabling the process to be automated. As long as there's a human in the loop, it's going to be pretty easy to tell whether the command needs to be retried. But in something like a CI environment, where you may want to create and destroy clusters in an automated fashion, and any flakiness there would be irritating. 

But I'm not sure that errors of this nature are going to be common enough to be worth worrying about, so it's probably simpler to remove this part about counting nodes and wait and see if it becomes a problem. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14251)*

**Comment by mberhault** [comment]
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[docs/RFCS/init_command.md, line 109 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14251#-Kf_yzD6fEWGGkLNAUe-:-KgE5_zI91ZEeV5-0mcP:b-a18gh6) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab226028f90995c1e32f628b7636688f54ac5cc/docs/RFCS/init_command.md#L109)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Yeah, I was thinking that even with node-counting, we would require that the `init` command be issued directly to a node. I don't want to build in a forwarding mechanism (at this time). 

The point of counting nodes is to eliminate ambiguous results, enabling the process to be automated. As long as there's a human in the loop, it's going to be pretty easy to tell whether the command needs to be retried. But in something like a CI environment, where you may want to create and destroy clusters in an automated fashion, and any flakiness there would be irritating. 

But I'm not sure that errors of this nature are going to be common enough to be worth worrying about, so it's probably simpler to remove this part about counting nodes and wait and see if it becomes a problem. 
</blockquote></details>

ok, if you make it a requirement that we can access nodes directly, then let's skip the forwarding for now (maybe mention it in a "future improvements" section).
And as long as the target node is guaranteed to stay the same, I think we can skip counting. But it's easy enough to add, so if you want to keep it for either safety or clarify, feel free.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14251)*

**Comment by petermattis** [comment]
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[docs/RFCS/init_command.md, line 67 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14251#-KgFfjS-NeoaQcY-vzS-:-KgFfjS0x2jcXCVMwrgs:bnbelzq) ([raw file](https://github.com/cockroachdb/cockroach/blob/2abd445e0f26352ab33fb2908c4799595ac11ff7/docs/RFCS/init_command.md#L67)):*
> ```Markdown
> (lexicographically smallest address?) and calls `InitCluster` on it.
> If it fails for any reason other than the cluster already being
> bootstrapped, back off and try again.
> ```

During the back off and retry, will the same cluster node be used? I think this should be made explicit. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14251)*

**Comment by mberhault** [comment]
LGTM (after the last few comments. some work can be moved to a "future improvements" section)

**Comment by jseldess** [comment]
*[docs/RFCS/init_command.md, line 20 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14251#-KfXzSJJXwJEvWhtxZxd-r2-20:-Kgyy4DGWpnG1xG3-5K_:b-8syfyf) ([raw file](https://github.com/cockroachdb/cockroach/blob/2abd445e0f26352ab33fb2908c4799595ac11ff7/docs/RFCS/init_command.md#L20)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

An example using a three node cluster would be nice here.
</blockquote></details>

+1   

It'd be great to add 2 examples, actually, one for a local test cluster and another for a real distributed cluster.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14251)*

**Comment by sonix07** [comment]
would it be possible by using the init command to have a docker swarm service with multiple replicas instead of creating a different service for each cockroachdb instance?https://github.com/cockroachdb/cockroach/issues/17133

**Comment by a-robinson** [comment]
It would make things much less awkward, yeah. That's part of the reason it's being built :) Let's try to move the discussion back to #17133, though.
